connor/poolModule.go

```diff
@@ -0,0 +1,248 @@
+
+			log.Printf("Iteration for worker :: %v more than 4 == > get Avg Data", w.DealID)
+			changeAvgWorker := 100 - ((uint64(w.WorkerAvgHashrate*hashes) * 100) / bidHashrate)
+			if err = p.DetectingDeviation(ctx, changeAvgWorker, w, dealInfo); err != nil {
+				return err
+			}
+		}
+		p.c.db.UpdateIterationPoolDB(w.DealID, iteration)
+	}
+	return nil
+}
+
+//Detection of getting a lowered hashrate and sending to a blacklist (create deal finish request).
+func (p *PoolModule) DetectingDeviation(ctx context.Context, changePercentDeviationWorker uint64, worker *database.PoolDb, dealInfo *sonm.DealInfoReply) error {
+	if changePercentDeviationWorker >= uint64(p.c.cfg.Sensitivity.WorkerLimitChangePercent) {
+		if worker.BadGuy < numberOfLives {
+			newStatus := worker.BadGuy + 1
+			p.c.db.UpdateWorkerStatusInPoolDB(worker.DealID, newStatus, time.Now())
+		} else {
+			if err := p.DestroyDeal(ctx, dealInfo); err != nil {
+				return err
+			}
+			p.c.db.UpdateWorkerStatusInPoolDB(worker.DealID, int32(BanStatusWorkerInPool), time.Now())
+			log.Printf("This deal is destroyed (bad status in Pool) : %v!\r\n", dealInfo.Deal.Id)
+		}
+	} else if changePercentDeviationWorker >= 20 {
+		p.DestroyDeal(ctx, dealInfo)
+		p.c.db.UpdateWorkerStatusInPoolDB(worker.DealID, int32(BanStatusWorkerInPool), time.Now())
+		log.Printf("This deal is destroyed (Pidor more than 5) : %v!\r\n", dealInfo.Deal.Id)
+	}
+	return nil
+}
+
+// Update pool data for first hour (use reported hashrate without shares)
+func (p *PoolModule) UpdateRHPoolData(ctx context.Context, poolRHData watchers.PoolWatcher, addr string) error {
+	poolRHData.Update(ctx)
+	dataRH, err := poolRHData.GetData(addr)
+	if err != nil {
+		log.Printf("Cannot get data RH  --> %v\r\n", err)
+		return err
+	}
+
+	for _, rh := range dataRH.Data {
+		p.c.db.UpdateReportedHashratePoolDB(rh.Worker, rh.Hashrate, time.Now())
+		log.Printf("Update RH in DB %v, %v, %v\r\n", rh.Worker, rh.Hashrate, time.Now())
+	}
+	return nil
+}
+
+// Update pool data for another time (use average hashrate with shares)
+func (p *PoolModule) UpdateAvgPoolData(ctx context.Context, poolAvgData watchers.PoolWatcher, addr string) error {
+	poolAvgData.Update(ctx)
+	dataRH, err := poolAvgData.GetData(addr)
+	if err != nil {
+		log.Printf("Cannot get data AvgPool  --> %v\r\n", err)
+		return err
+	}
+
+	for _, rh := range dataRH.Data {
+		p.c.db.UpdateAvgPoolDB(rh.Worker, rh.Hashrate, time.Now())
+	}
+	return nil
+}
+
+// Send to Connor's blacklist failed worker. If percent of failed workers more than "cleaner" workers => send Master to blacklist and destroy deal.
+func (p *PoolModule) SendToConnorBlackList(ctx context.Context, failedDeal *sonm.DealInfoReply) error {
+	workerList, err := p.c.MasterClient.WorkersList(ctx, failedDeal.Deal.MasterID)
+	if err != nil {
+		return err
+	}
+
+	for _, wM := range workerList.Workers {
+		val, err := p.c.db.GetBlacklistFromDb(wM.SlaveID.String())
```

Comment:
here you calling a `.String` method provided by the protobuf, use an EthAddress.Unwrap().Hex() method if you want to represent this as valid eth address. 

UPD: I think you should replace it everywhere.